In [1]:
import pandas as pd

In [3]:
# load the dataset as a DataFrame
data = pd.read_csv("data.csv")

In [4]:
# Get an overview of the data
data.head()

play_id     game_id home_team away_team posteam posteam_type defteam  \
0       46  2009091000       PIT       TEN     PIT         home     TEN   
1       68  2009091000       PIT       TEN     PIT         home     TEN   
2       92  2009091000       PIT       TEN     PIT         home     TEN   
3      113  2009091000       PIT       TEN     PIT         home     TEN   
4      139  2009091000       PIT       TEN     PIT         home     TEN   

  side_of_field  yardline_100   game_date  ...  penalty_player_id  \
0           TEN          30.0  2009-09-10  ...                NaN   
1           PIT          58.0  2009-09-10  ...                NaN   
2           PIT          53.0  2009-09-10  ...                NaN   
3           PIT          56.0  2009-09-10  ...                NaN   
4           PIT          56.0  2009-09-10  ...                NaN   

   penalty_player_name  penalty_yards replay_or_challenge  \
0                  NaN            NaN                   0   
1                  NaN            NaN                   0   
2                  NaN            NaN                   0   
3                  NaN            NaN                   0   
4                  NaN            NaN                   0   

   replay_or_challenge_result  penalty_type  defensive_two_point_attempt  \
0                         NaN           NaN                          0.0   
1                         NaN           NaN                          0.0   
2                         NaN           NaN                          0.0   
3                         NaN           NaN                          0.0   
4                         NaN           NaN                          0.0   

   defensive_two_point_conv  defensive_extra_point_attempt  \
0                       0.0                            0.0   
1                       0.0                            0.0   
2                       0.0                            0.0   
3                       0.0                            0.0   
4                       0.0                            0.0   

   defensive_extra_point_conv  
0                         0.0  
1                         0.0  
2                         0.0  
3                         0.0  
4                         0.0  

[5 rows x 256 columns]

- The data is about players, team and there games..
- There is a category and numaric columns
- There is NAN value need to handling
- There is date column
- There is column has 0/1 value -> its category not numaric


In [5]:
# Get the shape of the data 
data.shape

(44596, 256)

we have 44596 rows and 256 column to handle

In [6]:
# Delete columns that were missing data more than 65% of the data of the column 
data.drop(data.columns[data.isna().mean() > 0.65], axis=1, inplace=True)

In [7]:
# Display the shape of data DataFrame after deleting
data.shape

(44596, 165)

We had 256 columns and now have 165 columns, which means that 91 columns had missing data that represented for more than 65 percent of the column's data.

In [8]:
# I will eliminate rows when more than half of the row is missing.
data.drop(data.index[data.isna().mean(axis=1) >= 0.5].tolist(),inplace=True)

In [9]:
# Display the shape of numeric_data DataFrame after deleting
data.shape

(43084, 165)

decreased the number of rows by around 1512 rows

In [10]:
# What is the data type that data contains?
data.dtypes.unique()

array([dtype('int64'), dtype('O'), dtype('float64')], dtype=object)

I had to split the data into numeric data that contain int64 and float64 data type, and category data that contain object data type because handling the numeric data is not the same as categorical data. 

In [11]:
numeric_data = data.select_dtypes(exclude=['object']).copy()
print("numeric shape:", numeric_data.shape)
category_data = data.select_dtypes(exclude=['int64', 'float64']).copy()
print("category shape:", category_data.shape)

numeric shape: (43084, 144)
category shape: (43084, 21)


# Handling Numeric Data

In [12]:
# Let's see how many columns I want to deal with to handle it
numeric_data.columns[numeric_data.isna().sum() > 0].shape

(36,)

36 out of 144 I will handle them

In [13]:
# Get features name which has NAN value from numeric columns of mean_features columns
mean_features = numeric_data.columns[numeric_data.isna().any()]

In [14]:
mean_features.shape 

(36,)

There are 36 columns to deal with, all of which have missing data

In [15]:
# Fill NAN value for these columns by the column mean
numeric_data[mean_features] = numeric_data[mean_features].fillna(value=numeric_data[mean_features].mean())

In [16]:
# Check to see if any columns have missing data
numeric_data.columns[numeric_data.isna().any()]

Index([], dtype='object')

We've finished with the numerical data, now I'll enter it into the original dataframe

# Handling Category Data

In [17]:
category_data.head()

home_team away_team posteam posteam_type defteam side_of_field   game_date  \
0       PIT       TEN     PIT         home     TEN           TEN  2009-09-10   
1       PIT       TEN     PIT         home     TEN           PIT  2009-09-10   
2       PIT       TEN     PIT         home     TEN           PIT  2009-09-10   
3       PIT       TEN     PIT         home     TEN           PIT  2009-09-10   
4       PIT       TEN     PIT         home     TEN           PIT  2009-09-10   

  game_half   time   yrdln  ... play_type pass_length pass_location  \
0     Half1  15:00  TEN 30  ...   kickoff         NaN           NaN   
1     Half1  14:53  PIT 42  ...      pass       short          left   
2     Half1  14:16  PIT 47  ...       run         NaN           NaN   
3     Half1  13:35  PIT 44  ...      pass        deep         right   
4     Half1  13:27  PIT 44  ...      punt         NaN           NaN   

  passer_player_id passer_player_name receiver_player_id receiver_player_name  \
0              NaN                NaN                NaN                  NaN   
1       00-0022924   B.Roethlisberger         00-0017162               H.Ward   
2              NaN                NaN                NaN                  NaN   
3       00-0022924   B.Roethlisberger         00-0026901            M.Wallace   
4              NaN                NaN                NaN                  NaN   

  solo_tackle_1_team solo_tackle_1_player_id solo_tackle_1_player_name  
0                TEN              00-0025406                 M.Griffin  
1                TEN              00-0021219                    C.Hope  
2                TEN              00-0024331                 S.Tulloch  
3                NaN                     NaN                       NaN  
4                NaN                     NaN                       NaN  

[5 rows x 21 columns]

In [18]:
# Whay is the number of columns that has missing data ?
category_data.columns[category_data.isna().any()].shape

(14,)

14 out of 21 I will handle them

In [19]:
# Display the columns to determine how each one should be handled
category_data.columns[category_data.isna().any()]

Index(['posteam', 'posteam_type', 'defteam', 'side_of_field', 'yrdln',
       'pass_length', 'pass_location', 'passer_player_id',
       'passer_player_name', 'receiver_player_id', 'receiver_player_name',
       'solo_tackle_1_team', 'solo_tackle_1_player_id',
       'solo_tackle_1_player_name'],
      dtype='object')

In [20]:
category_data.columns[category_data.isna().any()][0:]

Index(['posteam', 'posteam_type', 'defteam', 'side_of_field', 'yrdln',
       'pass_length', 'pass_location', 'passer_player_id',
       'passer_player_name', 'receiver_player_id', 'receiver_player_name',
       'solo_tackle_1_team', 'solo_tackle_1_player_id',
       'solo_tackle_1_player_name'],
      dtype='object')

We can handle these columns using the back and farward fill technique:
    
    ['passer_player_id', 'passer_player_name', 'receiver_player_id', 'receiver_player_name', 'solo_tackle_1_player_id', 'solo_tackle_1_player_name']


We can handle these columns based on the most frequent value:
    
    ['posteam', 'posteam_type', 'defteam', 'side_of_field', 'yrdln', 'pass_length', 'pass_location', 'solo_tackle_1_team']

In [21]:
# Splits the columns and save their name into lists
back_fareard_fill_columns = category_data.columns[category_data.isna().any()][7:].drop('solo_tackle_1_team')
most_frequent_fill_columns = category_data.columns[category_data.isna().any()].drop(back_fareard_fill_columns)

In [22]:
# Fill missing data by farward method
category_data[back_fareard_fill_columns] = category_data[back_fareard_fill_columns].fillna(method='ffill')

In [23]:
# Fill missing data by back method
category_data[back_fareard_fill_columns] = category_data[back_fareard_fill_columns].fillna(method='bfill')

In [24]:
# Check the missing data for back_fareard_fill_columns 
category_data[back_fareard_fill_columns].isna().any()

passer_player_id             False
passer_player_name           False
receiver_player_id           False
receiver_player_name         False
solo_tackle_1_player_id      False
solo_tackle_1_player_name    False
dtype: bool

In [25]:
# Fill missing data with most frequent value for each column in most_frequent_fill_columns
category_data[most_frequent_fill_columns] = category_data[most_frequent_fill_columns].fillna(category_data[most_frequent_fill_columns].mode().iloc[0])

In [26]:
# Check missing data for the category dataframe
category_data.isna().any()

home_team                    False
away_team                    False
posteam                      False
posteam_type                 False
defteam                      False
side_of_field                False
game_date                    False
game_half                    False
time                         False
yrdln                        False
desc                         False
play_type                    False
pass_length                  False
pass_location                False
passer_player_id             False
passer_player_name           False
receiver_player_id           False
receiver_player_name         False
solo_tackle_1_team           False
solo_tackle_1_player_id      False
solo_tackle_1_player_name    False
dtype: bool

Everything is right, now we will merge all them to get the final datafram

# Get the final dataframe

In [27]:
data = numeric_data.join(category_data)

In [28]:
data.head(10)

play_id     game_id  yardline_100  quarter_seconds_remaining  \
0       46  2009091000          30.0                      900.0   
1       68  2009091000          58.0                      893.0   
2       92  2009091000          53.0                      856.0   
3      113  2009091000          56.0                      815.0   
4      139  2009091000          56.0                      807.0   
5      162  2009091000          98.0                      796.0   
6      183  2009091000          98.0                      760.0   
7      207  2009091000          94.0                      731.0   
8      228  2009091000          96.0                      694.0   
9      253  2009091000          43.0                      684.0   

   half_seconds_remaining  game_seconds_remaining  quarter_end  drive  sp  \
0                  1800.0                  3600.0            0      1   0   
1                  1793.0                  3593.0            0      1   0   
2                  1756.0                  3556.0            0      1   0   
3                  1715.0                  3515.0            0      1   0   
4                  1707.0                  3507.0            0      1   0   
5                  1696.0                  3496.0            0      2   0   
6                  1660.0                  3460.0            0      2   0   
7                  1631.0                  3431.0            0      2   0   
8                  1594.0                  3394.0            0      2   0   
9                  1584.0                  3384.0            0      3   0   

   qtr  ...  play_type  pass_length  pass_location  passer_player_id  \
0    1  ...    kickoff        short          right        00-0022924   
1    1  ...       pass        short           left        00-0022924   
2    1  ...        run        short          right        00-0022924   
3    1  ...       pass         deep          right        00-0022924   
4    1  ...       punt        short          right        00-0022924   
5    1  ...        run        short          right        00-0022924   
6    1  ...       pass        short           left        00-0003292   
7    1  ...        run        short          right        00-0003292   
8    1  ...       punt        short          right        00-0003292   
9    1  ...       pass        short          right        00-0022924   

   passer_player_name  receiver_player_id  receiver_player_name  \
0    B.Roethlisberger          00-0017162                H.Ward   
1    B.Roethlisberger          00-0017162                H.Ward   
2    B.Roethlisberger          00-0017162                H.Ward   
3    B.Roethlisberger          00-0026901             M.Wallace   
4    B.Roethlisberger          00-0026901             M.Wallace   
5    B.Roethlisberger          00-0026901             M.Wallace   
6           K.Collins          00-0024489                A.Hall   
7           K.Collins          00-0024489                A.Hall   
8           K.Collins          00-0024489                A.Hall   
9    B.Roethlisberger          00-0026901             M.Wallace   

   solo_tackle_1_team  solo_tackle_1_player_id  solo_tackle_1_player_name  
0                 TEN               00-0025406                  M.Griffin  
1                 TEN               00-0021219                     C.Hope  
2                 TEN               00-0024331                  S.Tulloch  
3                 STL               00-0024331                  S.Tulloch  
4                 STL               00-0024331                  S.Tulloch  
5                 STL               00-0024331                  S.Tulloch  
6                 PIT               00-0022119                 T.Polamalu  
7                 PIT               00-0022119                 T.Polamalu  
8                 TEN               00-0025406                  M.Griffin  
9                 TEN               00-0026243                    W.Hayes  

[10 rows x 165 columns]

Finally, I wish you a wonderful day! :)